File: iOS/Catalyst/Sources/Catalyst/Tokens/CatColorPalette.swift
//
//  CatColorPalette.swift
//  Catalyst
//
//  Created by Efe Durmaz on 06.13.26.
//

import SwiftUI

// MARK: - CatColorPalette

/// A concrete bag of the 9 semantic color slots that every `CatColors.Theme.*` group exposes.
///
/// Consumers read `palette.bg`, `palette.fill`, etc. without caring which token group or
/// theme produced the values. Adding a new theme only requires a new entry in `CatColorPalette.registry`
/// — no extensions, no per-component changes.
public struct CatColorPalette {
    public let bg: Color
    public let bgHover: Color
    public let bgActive: Color
    public let fill: Color
    public let fillHover: Color
    public let fillActive: Color
    public let text: Color
    public let textHover: Color
    public let textActive: Color
}

// MARK: - Registry

extension CatColorPalette {

    /// Central lookup table: `[ThemeType: [ColorConfig: CatColorPalette]]`.
    ///
    /// To add a new theme (e.g. `.dark`), append one block here pointing at the new
    /// `CatColors.DarkTheme.*` token groups. Nothing else in the codebase needs to change.
    ///
    /// ```swift
    /// // Adding dark theme — one new entry in this dictionary:
    /// .dark: [
    ///     .primary: .init(CatColors.DarkTheme.Primary.self),
    ///     .danger:  .init(CatColors.DarkTheme.Danger.self),
    ///     // ...
    /// ]
    /// ```
    static let registry: [CatTheme.ThemeType: [CatTheme.ColorConfig: CatColorPalette]] = [
        .primaryHaiilo: [
            .danger: .init(CatColors.Theme.Danger.self),
            .info: .init(CatColors.Theme.Info.self),
            .primary: .init(CatColors.Theme.Primary.self),
            .primaryInverted: .init(CatColors.Theme.PrimaryInverted.self),
            .secondary: .init(CatColors.Theme.Secondary.self),
            .secondaryInverted: .init(CatColors.Theme.SecondaryInverted.self),
            .success: .init(CatColors.Theme.Success.self),
            .warning: .init(CatColors.Theme.Warning.self)
        ]
    ]
}

// MARK: - CatThemeColorTokens protocol

/// Anything that provides the 9 standard semantic color slots.
/// Each `CatColors.Theme.*` enum conforms to this automatically via the extension below —
/// no manual conformance needed when new groups are added to the generated file.
public protocol CatThemeColorTokens {
    static var bg: Color { get }
    static var bgHover: Color { get }
    static var bgActive: Color { get }
    static var fill: Color { get }
    static var fillHover: Color { get }
    static var fillActive: Color { get }
    static var text: Color { get }
    static var textHover: Color { get }
    static var textActive: Color { get }
}

extension CatColors.Theme.Danger: CatThemeColorTokens {}
extension CatColors.Theme.Info: CatThemeColorTokens {}
extension CatColors.Theme.Primary: CatThemeColorTokens {}
extension CatColors.Theme.PrimaryInverted: CatThemeColorTokens {}
extension CatColors.Theme.Secondary: CatThemeColorTokens {}
extension CatColors.Theme.SecondaryInverted: CatThemeColorTokens {}
extension CatColors.Theme.Success: CatThemeColorTokens {}
extension CatColors.Theme.Warning: CatThemeColorTokens {}

extension CatColorPalette {
    /// Lifts any `CatThemeColorTokens`-conforming type into a `CatColorPalette` value.
    /// Used inside `registry` so each entry is a one-liner: `.init(CatColors.Theme.Danger.self)`.
    init<T: CatThemeColorTokens>(_ tokens: T.Type) {
        self.init(
            bg: T.bg,
            bgHover: T.bgHover,
            bgActive: T.bgActive,
            fill: T.fill,
            fillHover: T.fillHover,
            fillActive: T.fillActive,
            text: T.text,
            textHover: T.textHover,
            textActive: T.textActive
        )
    }

    /// Derives all 9 palette slots from a single arbitrary accent `Color`.
    ///
    /// This powers the whitelabeling path: clients pass one `Color` via
    /// `.catalystAccentColor(_:)` and every button variant (`.filled`, `.outlined`,
    /// `.text`, `.link`) automatically picks up correct hover/pressed states.
    ///
    /// - `bg` slots — used by the `.filled` variant background. Darkened on interaction
    ///   using `AccentColorDarkenFactor`.
    /// - `fill` slots — foreground *inside* a filled button. Kept white because accent
    ///   backgrounds are assumed to be dark enough to require white text (mirrors the
    ///   existing `Components.Buttons.Accent.filledConfig` behavior).
    /// - `text` slots — used by `.outlined`, `.text`, and `.link` variant foregrounds.
    ///   Normal state uses the accent color directly; hover/pressed darken it.
    init(accentColor: Color) {
        let hoverFactor  = CatTheme.AccentColorDarkenFactor.hovered.rawValue
        let activeFactor = CatTheme.AccentColorDarkenFactor.pressed.rawValue
        self.init(
            bg:          accentColor,
            bgHover:     accentColor.darken(by: hoverFactor),
            bgActive:    accentColor.darken(by: activeFactor),
            fill:        Color.white,
            fillHover:   Color.white,
            fillActive:  Color.white,
            text:        accentColor,
            textHover:   accentColor.darken(by: hoverFactor),
            textActive:  accentColor.darken(by: activeFactor)
        )
    }
}

File: android/catalyst/src/main/java/com/haiilo/catalyst/theme/CatColorPalette.kt
package com.haiilo.catalyst.theme

import androidx.compose.ui.graphics.Color
import com.haiilo.catalyst.extensions.darken
import com.haiilo.catalyst.tokens.generated.CatColors

// ---------------------------------------------------------------------------
// CatColorPalette
//
// A resolved bag of the 9 semantic color slots exposed by every
// CatColors.Theme.* token group. Components read palette.bg, palette.text,
// etc. without coupling themselves to a specific token object or color role.
//
// Usage:
//
//   val p = CatColorPalette.Primary
//   val p = CatColorPalette.Danger
//
// or build a fully custom palette for whitelabeling:
//
//   val p = CatColorPalette(bg = myColor, ...)
// ---------------------------------------------------------------------------

data class CatColorPalette(
    val bg: Color,
    val bgHover: Color,
    val bgActive: Color,
    val fill: Color,
    val fillHover: Color,
    val fillActive: Color,
    val text: Color,
    val textHover: Color,
    val textActive: Color,
) {
    companion object {
        // -------------------------------------------------------------------
        // Accent color support
        // -------------------------------------------------------------------

        /**
         * Lightness-reduction factors applied to the accent base color to
         * produce hover and pressed variants.
         *
         * Mirrors `CatTheme.AccentColorDarkenFactor` on iOS.
         */
        object AccentDarkenFactor {
            const val HOVERED = 0.05f
            const val PRESSED = 0.11f
        }

        /**
         * Builds a [CatColorPalette] driven by a single brand accent [color].
         *
         * The hover and active shades are derived by darkening the base color
         * using HSL lightness reduction — matching the iOS implementation.
         *
         * Only the `bg*`, `fill*`, and `text*` slots relevant to the
         * [CatButtonVariant.Filled] / Outlined / Text / Link variants are
         * populated here; white is used for the foreground fill so the accent
         * background always has legible contrast.
         *
         * Mirrors `CatColorPalette(accentColor:)` on iOS.
         */
        fun fromAccentColor(color: Color): CatColorPalette {
            val bgHover = color.darken(AccentDarkenFactor.HOVERED)
            val bgActive = color.darken(AccentDarkenFactor.PRESSED)
            return CatColorPalette(
                bg = color,
                bgHover = bgHover,
                bgActive = bgActive,
                fill = Color.White,
                fillHover = Color.White,
                fillActive = Color.White,
                text = color,
                textHover = bgHover,
                textActive = bgActive,
            )
        }

        val Primary: CatColorPalette
            get() = CatColorPalette(
                bg = CatColors.Theme.Primary.bg,
                bgHover = CatColors.Theme.Primary.bgHover,
                bgActive = CatColors.Theme.Primary.bgActive,
                fill = CatColors.Theme.Primary.fill,
                fillHover = CatColors.Theme.Primary.fillHover,
                fillActive = CatColors.Theme.Primary.fillActive,
                text = CatColors.Theme.Primary.text,
                textHover = CatColors.Theme.Primary.textHover,
                textActive = CatColors.Theme.Primary.textActive,
            )

        val PrimaryInverted: CatColorPalette
            get() = CatColorPalette(
                bg = CatColors.Theme.PrimaryInverted.bg,
                bgHover = CatColors.Theme.PrimaryInverted.bgHover,
                bgActive = CatColors.Theme.PrimaryInverted.bgActive,
                fill = CatColors.Theme.PrimaryInverted.fill,
                fillHover = CatColors.Theme.PrimaryInverted.fillHover,
                fillActive = CatColors.Theme.PrimaryInverted.fillActive,
                text = CatColors.Theme.PrimaryInverted.text,
                textHover = CatColors.Theme.PrimaryInverted.textHover,
                textActive = CatColors.Theme.PrimaryInverted.textActive,
            )

        val Secondary: CatColorPalette
            get() = CatColorPalette(
                bg = CatColors.Theme.Secondary.bg,
                bgHover = CatColors.Theme.Secondary.bgHover,
                bgActive = CatColors.Theme.Secondary.bgActive,
                fill = CatColors.Theme.Secondary.fill,
                fillHover = CatColors.Theme.Secondary.fillHover,
                fillActive = CatColors.Theme.Secondary.fillActive,
                text = CatColors.Theme.Secondary.text,
                textHover = CatColors.Theme.Secondary.textHover,
                textActive = CatColors.Theme.Secondary.textActive,
            )

        val SecondaryInverted: CatColorPalette
            get() = CatColorPalette(
                bg = CatColors.Theme.SecondaryInverted.bg,
                bgHover = CatColors.Theme.SecondaryInverted.bgHover,
                bgActive = CatColors.Theme.SecondaryInverted.bgActive,
                fill = CatColors.Theme.SecondaryInverted.fill,
                fillHover = CatColors.Theme.SecondaryInverted.fillHover,
                fillActive = CatColors.Theme.SecondaryInverted.fillActive,
                text = CatColors.Theme.SecondaryInverted.text,
                textHover = CatColors.Theme.SecondaryInverted.textHover,
                textActive = CatColors.Theme.SecondaryInverted.textActive,
            )

        val Danger: CatColorPalette
            get() = CatColorPalette(
                bg = CatColors.Theme.Danger.bg,
                bgHover = CatColors.Theme.Danger.bgHover,
                bgActive = CatColors.Theme.Danger.bgActive,
                fill = CatColors.Theme.Danger.fill,
                fillHover = CatColors.Theme.Danger.fillHover,
                fillActive = CatColors.Theme.Danger.fillActive,
                text = CatColors.Theme.Danger.text,
                textHover = CatColors.Theme.Danger.textHover,
                textActive = CatColors.Theme.Danger.textActive,
            )

        val Success: CatColorPalette
            get() = CatColorPalette(
                bg = CatColors.Theme.Success.bg,
                bgHover = CatColors.Theme.Success.bgHover,
                bgActive = CatColors.Theme.Success.bgActive,
                fill = CatColors.Theme.Success.fill,
                fillHover = CatColors.Theme.Success.fillHover,
                fillActive = CatColors.Theme.Success.fillActive,
                text = CatColors.Theme.Success.text,
                textHover = CatColors.Theme.Success.textHover,
                textActive = CatColors.Theme.Success.textActive,
            )

        val Warning: CatColorPalette
            get() = CatColorPalette(
                bg = CatColors.Theme.Warning.bg,
                bgHover = CatColors.Theme.Warning.bgHover,
                bgActive = CatColors.Theme.Warning.bgActive,
                fill = CatColors.Theme.Warning.fill,
                fillHover = CatColors.Theme.Warning.fillHover,
                fillActive = CatColors.Theme.Warning.fillActive,
                text = CatColors.Theme.Warning.text,
                textHover = CatColors.Theme.Warning.textHover,
                textActive = CatColors.Theme.Warning.textActive,
            )

        val Info: CatColorPalette
            get() = CatColorPalette(
                bg = CatColors.Theme.Info.bg,
                bgHover = CatColors.Theme.Info.bgHover,
                bgActive = CatColors.Theme.Info.bgActive,
                fill = CatColors.Theme.Info.fill,
                fillHover = CatColors.Theme.Info.fillHover,
                fillActive = CatColors.Theme.Info.fillActive,
                text = CatColors.Theme.Info.text,
                textHover = CatColors.Theme.Info.textHover,
                textActive = CatColors.Theme.Info.textActive,
            )
    }
}
